### PR TITLE
REF: share __getitem__ for Categorical/PandasArray/DTA/TDA/PA

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/core/arrays/categorical.py [hunk 1]
```diff
@@ -1882,17 +1882,11 @@ def __getitem__(self, key):
...
-            i = self._codes[key]
-            return self._box_func(i)
-
-        key = check_array_indexer(self, key)
-
-        result = self._codes[key]
-        if result.ndim > 1:
+        result = super().__getitem__(key)
+        if np.ndim(result) > 1:
+            result = result._ndarray
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:1887]: shouldn't this be self._from_backing_data(result) ?

Comment (jreback) [reply pandas/core/arrays/categorical.py:1887]: what hits this case?

**Review by jreback — CHANGES_REQUESTED** [review]
just a question really, nice de-dupes

File: pandas/core/arrays/categorical.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/categorical.py:1887]: > shouldn't this be self._from_backing_data(result) ?

Not until we support 2D categorical.

> what hits this case?

`cat[:, None]`

Comment (jreback) [reply pandas/core/arrays/categorical.py:1887]: oh i c we allow this for matplotlib compat IIRC

**Review by jreback — APPROVED** [review]

File: pandas/core/arrays/categorical.py [hunk 2]
```diff
@@ -1882,17 +1882,11 @@ def __getitem__(self, key):
...
-        if isinstance(key, (int, np.integer)):
-            i = self._codes[key]
-            return self._box_func(i)
-
-        key = check_array_indexer(self, key)
-
-        result = self._codes[key]
-        if result.ndim > 1:
+        result = super().__getitem__(key)
+        if np.ndim(result) > 1:
```
Comment (simonjayhawkins) [inline pandas/core/arrays/categorical.py:1886]: Is `np.ndim(result) > 1` generally slower than something like `not lib.is_scalar(result) and result.ndim > 1`?

result = 5
%timeit np.ndim(result) > 1
2.77 µs ± 48.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit not lib.is_scalar(result) and result.ndim > 1
154 ns ± 1.2 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


or is this not important?

File: pandas/core/arrays/_mixins.py [hunk 3]
```diff
@@ -168,3 +172,22 @@ def _validate_setitem_key(self, key):
...
+
+        key = self._validate_getitem_key(key)
+        result = self._ndarray[key]
+        if lib.is_scalar(result):
+            return self._box_func(result)
+
+        result = self._from_backing_data(result)
+        return result
+
+    def _validate_getitem_key(self, key):
```
Comment (simonjayhawkins) [inline pandas/core/arrays/_mixins.py:192]: this duplicates `_validate_setitem_key`. is this needed?

in what cases could one of a getitem and a setitem key be valid and other invalid?

File: pandas/core/arrays/categorical.py (hunk 2 above)
Comment (jreback) [reply pandas/core/arrays/categorical.py:1886]: wow that's neat

Comment (jbrockmendel) [reply pandas/core/arrays/categorical.py:1886]: > not lib.is_scalar(result) and result.ndim > 1?

The trouble here is if we have a tuple result.  I'll take a look at finding a more performant solution.

File: pandas/core/arrays/_mixins.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/arrays/_mixins.py:192]: DatetimeLikeArrayMixin has special logic here.  #36210 is about deprecating that

File: pandas/core/arrays/categorical.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/categorical.py:1886]: I did wonder about that. but pandas\tests\arrays\categorical and pandas\tests\extension\test_categorical.py didn't fail with this change. Do we have tests for this.

also is it possible to get a nested tuple result? np.ndim would give 2 in this case?

**Comment by jreback** [comment]
@jbrockmendel merge away if good here (likely address @simonjayhawkins points as followups?)

File: pandas/core/arrays/categorical.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/categorical.py:1886]: Looks like `getattr(result, "ndim", 1) > 1` outperforms both np.ndim and lib.is_scalar checks, and dont need to worry about tuples.  Will update.

> also is it possible to get a nested tuple result? np.ndim would give 2 in this case?

Yes.  Supporting list-like object-dtype is a massive PITA.

File: pandas/core/arrays/_mixins.py [hunk 4]
```diff
@@ -30,6 +31,9 @@ def _from_backing_data(self: _T, arr: np.ndarray) -> _T:
         """
         raise AbstractMethodError(self)
 
+    def _box_func(self, x):
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/_mixins.py:34]: Can you add a docstring for this function?

**Pull request merged**